### PR TITLE
MRG: Use hatchling as build backend

### PR DESCRIPTION
**Pull request by hoechenberger**
Also dynamically generate version number via hatch-vcs

Addresses #1174

cc @ofek

Merge checklist
---------------

Maintainer, please confirm the following before merging.
If applicable:

- [x] All comments are resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/main/doc/whats_new.rst) is updated

### CODE REVIEW

File: doc/whats_new.rst [hunk 1]
```diff
@@ -36,7 +36,9 @@ Detailed list of changes
 🧐 API and behavior changes
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-- nothing yet
+- The package build backend has been switched from ``setuptools`` to ``hatchling``. This
+  only affects users who build and install MNE-BIDS from source, and should not lead to
+  changed runtime behavior, by `Richard Höchenberger`_ (:gh:`1204`)
```
Comment (sappelhoff) [inline doc/whats_new.rst:41]: in what ways does it affect users who build and install mne-bids from source? :)

Comment (sappelhoff) [reply doc/whats_new.rst:41]: I am not sure it should be in the release notes then🤔 usually we skip "infrastructure changes without effect on the user", don't we?

File: pyproject.toml [hunk 2]
```diff
@@ -80,7 +74,7 @@ doc = [
     "mne-nirs @ https://github.com/mne-tools/mne-nirs/archive/refs/heads/main.zip",
     "seaborn",
     "openneuro-py",
-    "defusedxml",  # for reading BrainVision montages: `examples/convert_eeg_to_bids.py`
+    "defusedxml",                                                                   # for reading BrainVision montages: `examples/convert_eeg_to_bids.py`
```
Comment (sappelhoff) [inline pyproject.toml:77]: is this is mistake or by design? if by design, why?

**Review by sappelhoff — COMMENTED** [review]
in general, I think this is nice -- thanks for adding it! 

It would also need updates to the release protocol in the WIKI

File: doc/whats_new.rst (hunk 1 above)
Comment (hoechenberger) [reply doc/whats_new.rst:41]: There should be a Code Health section like we have in MNE-BIDS-Pipeline.

**Review by sappelhoff — APPROVED** [review]
+1 to merge if:

- the changelog entry is moved to a new, adequate "codehealth" section
- this is resolved: https://github.com/mne-tools/mne-bids/pull/1204/files#r1419479479

We can update the release protocol separately.

Thanks!!

File: doc/whats_new.rst (hunk 1 above)
Comment (hoechenberger) [reply doc/whats_new.rst:41]: MS does the same for the VS Code Python extension 

image

File: pyproject.toml (hunk 2 above)
Comment (hoechenberger) [reply pyproject.toml:77]: Btw the linter I use comes with the Even Better TOML extension for VS Code

Very helpful tool really

**Comment by hoechenberger** [comment]
Thanks
Will do later tonight

**Comment by hoechenberger** [comment]
@sappelhoff I think we're good to go

**Review by sappelhoff — APPROVED** [review]

File: doc/whats_new.rst [hunk 3]
```diff
@@ -51,6 +51,14 @@ Detailed list of changes
 
 - nothing yet
 
+⚕️ Code health
+^^^^^^^^^^^^^
```
[the remark this answers is not in the document]
Comment (sappelhoff) [reply doc/whats_new.rst]: image

**Pull request merged**